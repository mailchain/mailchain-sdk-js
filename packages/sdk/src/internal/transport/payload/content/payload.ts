import { PayloadHeaders } from './headers';

/**
 * Payload.
 */
export interface Payload {
	/**
	 * @see PayloadHeaders
	 */
	Headers: PayloadHeaders;
	/**
	 * Raw/data/object that are decrypted and parsed base on the the headers.
	 */
	Content: Buffer;
}

/**
 * Encrypted payload.
 */
export interface EncryptedPayload {
	EncryptedHeaders: Buffer;

	EncryptedContentChunks: Buffer[];
}
