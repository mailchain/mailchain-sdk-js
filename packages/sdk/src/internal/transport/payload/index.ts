export type { Payload } from './payload';
export { serializeAndEncryptPayload } from './payload';
