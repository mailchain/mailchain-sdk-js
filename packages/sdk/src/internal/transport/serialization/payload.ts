/**
 * Encrypted payload.
 */
export interface EncryptedPayload {
	EncryptedHeaders: Buffer;

	EncryptedContentChunks: Buffer[];
}
