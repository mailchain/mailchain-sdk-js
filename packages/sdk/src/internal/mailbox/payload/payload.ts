import { ReadonlyMailPayload } from '../../../receiving/mail';
import { ReadonlyMailerPayload } from '../../../transport';
import { MailPayloadHeaders } from './headers';

/**
 * Payload.
 */
export interface MailPayload {
	/**
	 * @see MailPayloadHeaders
	 */
	Headers: MailPayloadHeaders;
	/**
	 * Raw/data/object that are decrypted and parsed base on the the headers.
	 */
	Content: Buffer;
}

/**
 * Encrypted payload.
 */
export interface MailEncryptedPayload {
	EncryptedHeaders: Buffer;

	EncryptedContentChunks: Buffer[];
}

export function convertPayload(payload: ReadonlyMailPayload): MailPayload {
	switch (payload.Headers.ContentType) {
		case 'message/x.mailchain':
			return payload;
		case 'message/x.mailchain-mailer':
			const castPayload = payload as ReadonlyMailerPayload;

			const headers = {
				...payload.Headers,
				MailerContent: castPayload.MailerContent,
			} as MailPayloadHeaders;

			return {
				Headers: headers,
				Content: castPayload.Content,
			};
		default:
			throw new Error(`Unsupported content type: ${payload.Headers.ContentType}`);
	}
}
