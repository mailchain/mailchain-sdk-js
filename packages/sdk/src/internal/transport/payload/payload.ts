import { ED25519ExtendedPrivateKey } from '@mailchain/crypto';
import { CHUNK_LENGTH_1MB } from './chunk';
import { encryptPayload } from './encrypt';
import { PayloadHeaders } from './headers';
import { serialize } from './serialization';

/**
 * Payload.
 */
export interface Payload {
	/**
	 * @see PayloadHeaders
	 */
	Headers: PayloadHeaders;
	/**
	 * Raw/data/object that are decrypted and parsed base on the the headers.
	 */
	Content: Buffer;
}

/**
 * Encrypted payload.
 */
export interface EncryptedPayload {
	EncryptedHeaders: Buffer;

	EncryptedContentChunks: Buffer[];
}

export async function serializeAndEncryptPayload(
	payload: Payload,
	payloadRootEncryptionKey: ED25519ExtendedPrivateKey,
) {
	return serialize(await encryptPayload(payload, payloadRootEncryptionKey, CHUNK_LENGTH_1MB));
}
