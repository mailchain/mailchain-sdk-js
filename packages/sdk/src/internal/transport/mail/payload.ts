import { KindNaClSecretKey, SignerWithPublicKey } from '@mailchain/crypto';
import { EncodingTypes } from '@mailchain/encoding';
import { MailData, MailAddress } from '../../formatters/types';
import { Payload } from '../../transport/payload/content/payload';
import { createMimeMessage } from '../../formatters/generate';
import { LookupResult } from '../../identityKeys';

export type Distribution = {
	recipients: MailAddress[];
	payload: Payload;
};

export async function createMailPayloads(
	senderMessagingKey: SignerWithPublicKey,
	resolvedAddresses: Map<string, LookupResult>,
	mailData: MailData,
): Promise<{
	original: Payload;
	distributions: Distribution[];
}> {
	const message = await createMimeMessage(mailData, resolvedAddresses);

	const original = await createMailPayload(senderMessagingKey, message.original);
	const visibleRecipientsPayload = await createMailPayload(senderMessagingKey, message.visibleRecipients);
	const blindRecipients = await Promise.all(
		message.blindRecipients.map(async ({ recipient, content }) => ({
			recipients: [recipient],
			payload: await createMailPayload(senderMessagingKey, content),
		})),
	);

	return {
		original,
		distributions: [
			{
				recipients: [...mailData.recipients, ...mailData.carbonCopyRecipients],
				payload: visibleRecipientsPayload,
			},
			...blindRecipients,
		],
	};
}

export async function createMailPayload(
	senderMessagingKey: SignerWithPublicKey,
	contentPayload: string,
): Promise<Payload> {
	const contentBuffer = Buffer.from(contentPayload);
	return {
		Headers: {
			Origin: senderMessagingKey.publicKey,
			ContentSignature: await senderMessagingKey.sign(contentBuffer),
			Created: new Date(),
			ContentLength: contentBuffer.length,
			ContentType: 'message/x.mailchain',
			ContentEncoding: EncodingTypes.Base64,
			ContentEncryption: KindNaClSecretKey,
		},
		Content: contentBuffer,
	};
}
