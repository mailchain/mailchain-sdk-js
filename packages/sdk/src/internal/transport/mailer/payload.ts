import { KindNaClSecretKey, SignerWithPublicKey } from '@mailchain/crypto';
import { EncodingTypes } from '@mailchain/encoding';
import { Payload } from '../payload';
import { createContentBuffer, MailerContent } from './content';

export type ReadonlyMailerPayload = {
	MailerContent: MailerContent;
	Headers: {
		ContentType: 'message/x.mailchain-mailer';
	} & Payload['Headers'];
} & Payload;

export async function createMailerPayload(
	mailerMessagingKey: SignerWithPublicKey,
	contentPayload: MailerContent,
): Promise<Payload> {
	const contentBuffer = Buffer.from(createContentBuffer(contentPayload));
	return {
		Headers: {
			Origin: mailerMessagingKey.publicKey,
			ContentSignature: await mailerMessagingKey.sign(contentBuffer),
			Created: new Date(),
			ContentLength: contentBuffer.length,
			ContentType: 'message/x.mailchain-mailer',
			ContentEncoding: EncodingTypes.Base64,
			ContentEncryption: KindNaClSecretKey,
		},
		Content: contentBuffer,
	};
}
