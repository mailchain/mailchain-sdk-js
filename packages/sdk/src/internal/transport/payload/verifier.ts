import { Payload } from './payload';

export class ErrorPayloadSignatureInvalid extends Error {
	constructor() {
		super('payload signature invalid');
	}
}

export class PayloadOriginVerifier {
	static create() {
		return new PayloadOriginVerifier();
	}

	async verifyPayloadOrigin(payload: Payload) {
		if (!payload.Headers) {
			throw new Error('payload does not contain Headers');
		}

		if (!payload.Headers.Origin) {
			throw new Error('payload does not contain Headers.Origin');
		}

		if (!(await payload.Headers.Origin.verify(payload.Content, payload.Headers.ContentSignature))) {
			throw new ErrorPayloadSignatureInvalid();
		}
	}
}
