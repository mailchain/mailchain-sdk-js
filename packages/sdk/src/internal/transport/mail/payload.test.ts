import { AliceED25519PrivateKey } from '@mailchain/crypto/ed25519/test.const';
import { KeyRing } from '@mailchain/keyring';
import { dummyMailData, dummyMailDataResolvedAddresses } from '../../test.const';
import { createMailPayloads } from './payload';

const mockMimeMessageResult = {
	original: 'original',
	visibleRecipients: 'visible',
	blindRecipients: dummyMailData.blindCarbonCopyRecipients.map((recipient) => ({
		recipient,
		content: `content-${recipient.address}`,
	})),
};

const mockCreateMimeMessage = jest.fn();
jest.mock('@mailchain/sdk/internal/formatters/generate', () => ({
	createMimeMessage: (...params) => mockCreateMimeMessage(...params),
}));

describe('createMailPayloads', () => {
	let keyRing: KeyRing;

	beforeAll(() => {
		jest.useFakeTimers().setSystemTime(new Date('2022-06-06'));
	});

	beforeEach(() => {
		mockCreateMimeMessage.mockReturnValue(mockMimeMessageResult);
		keyRing = KeyRing.fromPrivateKey(AliceED25519PrivateKey);
	});

	it('create correct payload for sender of the message', async () => {
		const { original } = await createMailPayloads(
			keyRing.accountMessagingKey(),
			dummyMailDataResolvedAddresses,
			dummyMailData,
		);

		expect(mockCreateMimeMessage).toHaveBeenCalledWith(dummyMailData, dummyMailDataResolvedAddresses);
		expect(original).toMatchSnapshot('original');
		expect(original.Content.toString()).toEqual(mockMimeMessageResult.original);
	});

	it('create correct payload for visible recipients of the message', async () => {
		const { distributions } = await createMailPayloads(
			keyRing.accountMessagingKey(),
			dummyMailDataResolvedAddresses,
			dummyMailData,
		);

		expect(mockCreateMimeMessage).toHaveBeenCalledWith(dummyMailData, dummyMailDataResolvedAddresses);
		expect(distributions[0].payload).toMatchSnapshot('visible');
		expect(distributions[0].payload.Content.toString()).toEqual(mockMimeMessageResult.visibleRecipients);
		expect(distributions[0].recipients).toEqual([
			...dummyMailData.recipients,
			...dummyMailData.carbonCopyRecipients,
		]);
	});

	it('create correct payload for blinded recipients of the message', async () => {
		const { distributions } = await createMailPayloads(
			keyRing.accountMessagingKey(),
			dummyMailDataResolvedAddresses,
			dummyMailData,
		);

		expect(mockCreateMimeMessage).toHaveBeenCalledWith(dummyMailData, dummyMailDataResolvedAddresses);
		distributions.slice(1).forEach((bccParams, index) => {
			expect(bccParams.payload).toMatchSnapshot(`blinded ${dummyMailData.blindCarbonCopyRecipients[index].name}`);
			expect(bccParams.payload.Content.toString()).toEqual(mockMimeMessageResult.blindRecipients[index].content);
			expect(bccParams.recipients[0]).toEqual(dummyMailData.blindCarbonCopyRecipients[index]);
		});
	});
});
