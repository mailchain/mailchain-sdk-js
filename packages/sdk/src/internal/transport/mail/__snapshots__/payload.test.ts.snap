// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMailPayloads create correct payload for blinded recipients of the message: blinded bcc1 1`] = `
Object {
  "Content": Object {
    "data": Array [
      99,
      111,
      110,
      116,
      101,
      110,
      116,
      45,
      98,
      99,
      99,
      49,
      64,
      109,
      97,
      105,
      108,
      99,
      104,
      97,
      105,
      110,
      46,
      99,
      111,
      109,
    ],
    "type": "Buffer",
  },
  "Headers": Object {
    "ContentEncoding": "base64/plain",
    "ContentEncryption": "nacl-secret-key",
    "ContentLength": 26,
    "ContentSignature": Uint8Array [
      50,
      37,
      85,
      13,
      45,
      17,
      182,
      48,
      220,
      77,
      210,
      139,
      11,
      60,
      18,
      247,
      219,
      74,
      160,
      33,
      52,
      245,
      108,
      105,
      40,
      234,
      143,
      99,
      58,
      45,
      104,
      199,
      80,
      155,
      17,
      183,
      218,
      161,
      94,
      63,
      18,
      45,
      65,
      231,
      16,
      16,
      238,
      53,
      176,
      136,
      59,
      123,
      76,
      58,
      0,
      253,
      214,
      146,
      148,
      133,
      114,
      165,
      189,
      6,
    ],
    "ContentType": "message/x.mailchain",
    "Created": 2022-06-06T00:00:00.000Z,
    "Origin": ED25519PublicKey {
      "bytes": Uint8Array [
        215,
        190,
        70,
        167,
        83,
        44,
        32,
        26,
        113,
        228,
        19,
        133,
        8,
        32,
        97,
        216,
        16,
        67,
        139,
        65,
        185,
        41,
        147,
        108,
        27,
        108,
        75,
        48,
        31,
        33,
        50,
        155,
      ],
      "curve": "ed25519",
    },
  },
}
`;

exports[`createMailPayloads create correct payload for blinded recipients of the message: blinded bcc2 1`] = `
Object {
  "Content": Object {
    "data": Array [
      99,
      111,
      110,
      116,
      101,
      110,
      116,
      45,
      98,
      99,
      99,
      50,
      64,
      109,
      97,
      105,
      108,
      99,
      104,
      97,
      105,
      110,
      46,
      99,
      111,
      109,
    ],
    "type": "Buffer",
  },
  "Headers": Object {
    "ContentEncoding": "base64/plain",
    "ContentEncryption": "nacl-secret-key",
    "ContentLength": 26,
    "ContentSignature": Uint8Array [
      33,
      83,
      24,
      41,
      120,
      229,
      141,
      218,
      150,
      195,
      41,
      8,
      27,
      57,
      6,
      81,
      108,
      88,
      27,
      243,
      71,
      131,
      212,
      13,
      156,
      9,
      72,
      130,
      108,
      163,
      128,
      109,
      127,
      116,
      214,
      82,
      13,
      230,
      157,
      0,
      2,
      2,
      235,
      12,
      135,
      184,
      206,
      210,
      10,
      112,
      142,
      149,
      237,
      111,
      95,
      142,
      220,
      169,
      230,
      84,
      23,
      46,
      13,
      14,
    ],
    "ContentType": "message/x.mailchain",
    "Created": 2022-06-06T00:00:00.000Z,
    "Origin": ED25519PublicKey {
      "bytes": Uint8Array [
        215,
        190,
        70,
        167,
        83,
        44,
        32,
        26,
        113,
        228,
        19,
        133,
        8,
        32,
        97,
        216,
        16,
        67,
        139,
        65,
        185,
        41,
        147,
        108,
        27,
        108,
        75,
        48,
        31,
        33,
        50,
        155,
      ],
      "curve": "ed25519",
    },
  },
}
`;

exports[`createMailPayloads create correct payload for sender of the message: original 1`] = `
Object {
  "Content": Object {
    "data": Array [
      111,
      114,
      105,
      103,
      105,
      110,
      97,
      108,
    ],
    "type": "Buffer",
  },
  "Headers": Object {
    "ContentEncoding": "base64/plain",
    "ContentEncryption": "nacl-secret-key",
    "ContentLength": 8,
    "ContentSignature": Uint8Array [
      164,
      248,
      247,
      17,
      76,
      148,
      172,
      110,
      98,
      91,
      196,
      76,
      121,
      13,
      28,
      6,
      40,
      3,
      151,
      15,
      209,
      95,
      144,
      141,
      127,
      61,
      98,
      248,
      227,
      210,
      192,
      212,
      8,
      39,
      104,
      216,
      136,
      115,
      181,
      153,
      11,
      217,
      122,
      182,
      224,
      231,
      234,
      184,
      97,
      233,
      73,
      254,
      31,
      251,
      26,
      231,
      9,
      110,
      232,
      97,
      172,
      122,
      90,
      1,
    ],
    "ContentType": "message/x.mailchain",
    "Created": 2022-06-06T00:00:00.000Z,
    "Origin": ED25519PublicKey {
      "bytes": Uint8Array [
        215,
        190,
        70,
        167,
        83,
        44,
        32,
        26,
        113,
        228,
        19,
        133,
        8,
        32,
        97,
        216,
        16,
        67,
        139,
        65,
        185,
        41,
        147,
        108,
        27,
        108,
        75,
        48,
        31,
        33,
        50,
        155,
      ],
      "curve": "ed25519",
    },
  },
}
`;

exports[`createMailPayloads create correct payload for visible recipients of the message: visible 1`] = `
Object {
  "Content": Object {
    "data": Array [
      118,
      105,
      115,
      105,
      98,
      108,
      101,
    ],
    "type": "Buffer",
  },
  "Headers": Object {
    "ContentEncoding": "base64/plain",
    "ContentEncryption": "nacl-secret-key",
    "ContentLength": 7,
    "ContentSignature": Uint8Array [
      7,
      255,
      141,
      135,
      111,
      3,
      200,
      216,
      13,
      125,
      44,
      9,
      222,
      238,
      224,
      214,
      220,
      226,
      86,
      164,
      5,
      16,
      221,
      151,
      162,
      169,
      90,
      223,
      52,
      27,
      107,
      132,
      241,
      181,
      38,
      220,
      178,
      199,
      128,
      76,
      2,
      122,
      125,
      193,
      165,
      255,
      11,
      185,
      48,
      205,
      5,
      12,
      133,
      111,
      202,
      75,
      125,
      58,
      240,
      50,
      207,
      170,
      210,
      14,
    ],
    "ContentType": "message/x.mailchain",
    "Created": 2022-06-06T00:00:00.000Z,
    "Origin": ED25519PublicKey {
      "bytes": Uint8Array [
        215,
        190,
        70,
        167,
        83,
        44,
        32,
        26,
        113,
        228,
        19,
        133,
        8,
        32,
        97,
        216,
        16,
        67,
        139,
        65,
        185,
        41,
        147,
        108,
        27,
        108,
        75,
        48,
        31,
        33,
        50,
        155,
      ],
      "curve": "ed25519",
    },
  },
}
`;
